File: hard/Lc10.go
package hard

// "ab" "a*."
// 1.dp
//
//	dp[i][j] [i][j+1]
//	[i+1][j] [i+1][j+1]
//
// abcd  ab.*
//
//	i     j
//
// 转移方程 分类讨论 递归
// for i
//
//	for j
//
// |或
// if p[j] == * then dp[i][j] = dp[i][j-2] || dp[i-1][j] && match(i,j-1)
// if p[j] != * then dp[i][j] = dp[i-1][j-1] && match(i,j)
func isMatch(s string, p string) bool {
	m, n := len(s), len(p)
	dp := make([][]bool, m+1)
	for i, _ := range dp {
		dp[i] = make([]bool, n+1)
	}
	dp[0][0] = true
	for j := 2; j <= n; j++ {
		if p[j-1] == '*' {
			dp[0][j] = dp[0][j] || dp[0][j-2]
		}
	}
	match := func(i int, j int) bool {
		return s[i-1] == p[j-1] || p[j-1] == '.'
	}
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if p[j-1] == '*' {
				if j-2 >= 0 {
					dp[i][j] = dp[i][j-2] || (dp[i-1][j] && match(i, j-1))
				}
			} else {
				dp[i][j] = dp[i-1][j-1] && match(i, j)
			}
		}
	}
	return dp[m][n]
}

File: hard/Lc10/Lc10.py
"""
2022年3月30日
1 dp
2 mn
3 60min
"""
class Lc10:
    # 刷表法
    def isMatch(self, s: str, p: str) -> bool:
        ls, lp = len(s), len(p)
        dp = [[False for _ in range(lp+1)] for _ in range(ls+1)]
        # base case
        dp[0][0] = True
        for i in range(0, ls+1):
            for j in range(0, lp+1):
                if dp[i][j]:
                    if i+1 <= ls and j+1 <= lp and p[j] == '.':
                        dp[i+1][j+1] |= True
                    elif j+1 <= lp and p[j] == '*':
                        dp[i][j+1] |= True
                        # *可以匹配多个
                        k = i+1
                        while k <= ls and (p[j-1] == '.' or p[j-1] == s[k-1]):
                            dp[k][j+1] |= True
                            k += 1
                    elif i+1 <= ls and j+1 <= lp and p[j] == s[i]:
                        dp[i+1][j+1] |= True
                    if j+2 <= lp and p[j+1] == '*':
                        dp[i][j+2] |= True
        return dp[ls][lp]